### PR TITLE
feat(apig/env): add a resources to manage the environment for the shared APIG

### PR DESCRIPTION
**Pull request by Zippo-Wang**
**What this PR does / why we need it**:

Add a resources to manage the environment for the shared APIG.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
=== RUN   TestAccApiGatewayEnvironment_basic
=== PAUSE TestAccApiGatewayEnvironment_basic
=== CONT  TestAccApiGatewayEnvironment_basic
--- PASS: TestAccApiGatewayEnvironment_basic (26.17s)
PASS
```

### CODE REVIEW

File: docs/resources/api_gateway_environment.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+---
+subcategory: "API Gateway (Shared APIG)"
+---
+
+# huaweicloud_api_gateway_environment
+
+Manages an APIG environment resource within HuaweiCloud.
```
Comment (chengxiangdong) [inline docs/resources/api_gateway_environment.md]:
```suggestion
Manages a shared APIG environment resource within HuaweiCloud.
```

File: docs/resources/api_gateway_environment.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+* `region` - (Optional, String, ForceNew) Specifies the region where the instance is located.
+  If omitted, the provider-level region will be used. Changing this will create a new resource.
+
+* `name` - (Required, String) Specifies the environment name.
+  The valid length is limited from `3` to `64`, only letters, digits and underscores (_) are allowed.
+  The name must start with a letter.
+
+* `description` - (Optional, String) Specifies the environment description.
+  The value can contain a maximum of `255` characters, and the angle brackets (< and >) are not allowed.
```
Comment (chengxiangdong) [inline docs/resources/api_gateway_environment.md]:
```suggestion
  The value can contain a maximum of `255` characters.
```

File: docs/resources/api_gateway_environment.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  Chinese characters must be in **UTF-8** or **Unicode** format.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The ID of the dedicated environment.
```
Comment (chengxiangdong) [inline docs/resources/api_gateway_environment.md]:
```suggestion
* `id` - The environment ID.
```

File: huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	log.Printf("[DEBUG] env id is : %s", envId)
+	if err != nil {
+		return nil, fmt.Errorf("error creating APIG client %s", err)
+	}
+
+	envs, err := environments.List(client, environments.ListOpts{})
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("[DEBUG] envs is : %s", envs)
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go]:
```suggestion
	log.Printf("[DEBUG] List of shared APIG environments: %#v", envs)
```
HTTP request will print request, response and header details, usually we don't print the result.

File: docs/resources/api_gateway_environment.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+* `created_time` - The time when the environment was created.
```
Comment (chengxiangdong) [inline docs/resources/api_gateway_environment.md]:
```suggestion
* `create_time` - The time when the environment was created.
```

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	apigClient, err := cfg.ApiGatewayV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating API gateway client: %s", err)
+	}
+
+	n, err := GetEnvironment(apigClient, d.Get("name").(string), d.Id())
+	if err != nil {
+		return diag.Errorf("the environment is not exists: %s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error retrieving shared APIG environment")
```

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+	if d.HasChanges("name", "description") {
+		updateOpts := environments.UpdateOpts{
+			Name:        d.Get("name").(string),
+			Description: utils.String(d.Get("description").(string)),
+		}
+
+		_, err = environments.Update(apigClient, d.Id(), updateOpts)
+		if err != nil {
+			return diag.Errorf("error updating API gateway shared environment: %s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]:
```suggestion
			return diag.Errorf("error updating shared API gateway environment: %s", err)
```

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	apigClient, err := cfg.ApiGatewayV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating API gateway client: %s", err)
+	}
+
+	err = environments.Delete(apigClient, d.Id())
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error deleting API gateway shared environment")
```
Comment (chengxiangdong) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error deleting shared API gateway environment")
```

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+		},
+	}
+}
+
+func resourceAPIGatewayEnvironmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	apigClient, err := cfg.ApiGatewayV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating API gateway client: %s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]:
```suggestion
		return diag.Errorf("error creating shared API gateway client (v1) : %s", err)
```
Others also recommend printing this way.

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+	createOpts := environments.CreateOpts{
+		Name:        d.Get("name").(string),
+		Description: d.Get("description").(string),
+	}
+	log.Printf("[DEBUG] Create API gateway shared options: %s", utils.MarshalValue(createOpts))
+
+	rst, err := environments.Create(apigClient, createOpts)
+	if err != nil {
+		return diag.Errorf("error creating API gateway shared environment: %s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]:
```suggestion
		return diag.Errorf("error creating shared API gateway environment: %s", err)
```

File: docs/resources/api_gateway_environment.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  name        = "test"
+  description = "test env"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region where the instance is located.
```
Comment (Lance52259) [inline docs/resources/api_gateway_environment.md]:
```suggestion
* `region` - (Optional, String, ForceNew) Specifies the region where the environment is located.
```

Comment (Zippo-Wang) [reply docs/resources/api_gateway_environment.md]: I have fixed this problem

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	rst, err := environments.Create(apigClient, createOpts)
+	if err != nil {
+		return diag.Errorf("error creating shared API gateway environment: %s", err)
+	}
+	d.SetId(rst.Id)
+
+	return resourceAPIGatewayEnvironmentRead(ctx, d, meta)
+}
+
+func GetEnvironment(client *golangsdk.ServiceClient, envName, envID string) (*environments.Environment, error) {
```
Comment (Lance52259) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]:
```suggestion
func getEnvironment(client *golangsdk.ServiceClient, envName, envID string) (*environments.Environment, error) {
```

Comment (Zippo-Wang) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]: I have fixed this problem😊

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+	apigClient, err := cfg.ApiGatewayV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating shared API gateway client (v1) : %s", err)
+	}
+
+	createOpts := environments.CreateOpts{
+		Name:        d.Get("name").(string),
+		Description: d.Get("description").(string),
+	}
+	log.Printf("[DEBUG] Create API gateway shared options: %s", utils.MarshalValue(createOpts))
```
Comment (ShiChangkuo) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]:
```suggestion
	log.Printf("[DEBUG] Create API gateway shared options: %v", createOpts)
```

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"create_time": {
```
Comment (ShiChangkuo) [inline huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]: please define it as **created_at**

File: huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+func getEnvironmentFunc(cfg *config.Config, state *terraform.ResourceState) (interface{}, error) {
+	client, err := cfg.ApiGatewayV1Client(acceptance.HW_REGION_NAME)
+	envId := state.Primary.ID
+	log.Printf("[DEBUG] env id is : %s", envId)
+	if err != nil {
+		return nil, fmt.Errorf("error creating APIG client %s", err)
+	}
+
+	envs, err := environments.List(client, environments.ListOpts{})
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go]: it's better to filter environments by `name`.

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go (hunk 13 above)
Comment (wangzepeng666) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:64]: I wrote this like you before, but Xiangdong suggest me use this way to DEBUG. Can both methods be used?

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go (hunk 14 above)
Comment (wangzepeng666) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:44]: I use "create_time" due to it is written in the API document. Do we have to ues "create_at"?

File: huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go (hunk 15 above)
Comment (wangzepeng666) [reply huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go:26]: The id and name of environments are both the unique, so both of these can be used to determine an environment

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go (hunk 13 above)
Comment (Zippo-Wang) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]: > I wrote this like you before, but Xiangdong suggest me use this way to DEBUG. Can both methods be used?

I wrote this like you before, but Xiangdong suggest me use this way to DEBUG. Can both methods be used?

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go (hunk 14 above)
Comment (wangzepeng666) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:44]: > I use "create_time" due to it is written in the API document. Do we have to ues "create_at"?

I use "create_time" due to it is written in the API document. Do we have to ues "create_at"?

Comment (Zippo-Wang) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]: > please define it as **created_at**

I use "create_time" due to it is written in the API document. Do we have to ues "create_at"?

File: huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go (hunk 15 above)
Comment (Zippo-Wang) [reply huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go]: > it's better to filter environments by `name`.

The id and name of environments are both the unique, so both of these can be used to determine an environment

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go (hunk 14 above)
Comment (Zippo-Wang) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]: > please define it as **created_at**

I have fixed this problem, thanks for your valuable suggestions

File: huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go (hunk 15 above)
Comment (Zippo-Wang) [reply huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go]: > it's better to filter environments by `name`.

I have fixed this problem, thanks for your valuable suggestions

File: huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go (hunk 13 above)
Comment (Zippo-Wang) [reply huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go]: > 
I have fixed this problem, thanks for your valuable suggestions

**Pull request merged**